File: problems/hackerrank/matrix/organizingContainers/organizingContainers.go
package matrix
import (
	"sort"
)








// OrganizingContainers determines if it is possible to organize containers.
func OrganizingContainers(container [][]int) string {
	n := len(container)
	rowSums := make([]int, n)
	colSums := make([]int, n)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			rowSums[i] += container[i][j]
			colSums[j] += container[i][j]
		}
	}

	sort.Ints(rowSums)
	sort.Ints(colSums)

	for i := 0; i < n; i++ {
		if rowSums[i] != colSums[i] {
			return "Impossible"
		}
	}

	return "Possible"
}

// OrganizingContainersCounter uses frequency map approach.
func OrganizingContainersCounter(container [][]int) string {
	n := len(container)
	rowSums := make([]int, n)
	colSums := make([]int, n)

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			rowSums[i] += container[i][j]
			colSums[j] += container[i][j]
		}
	}

	rowCounts := make(map[int]int)
	colCounts := make(map[int]int)

	for i := 0; i < n; i++ {
		rowCounts[rowSums[i]]++
		colCounts[colSums[i]]++
	}

	if len(rowCounts) != len(colCounts) {
		return "Impossible"
	}

	for k, v := range rowCounts {
		if colCounts[k] != v {
			return "Impossible"
		}
	}

	return "Possible"
}

File: problems/hackerrank/matrix/organizingContainers/organizingContainers.py

from collections import Counter
import sys

#Submitted Solution on HackerRank
def organizingContainers(container):
    """
    Approach 1: The Sorting Method (Timsort)
    - Complexity: O(n log n)
    - Best for: Small to medium datasets (n < 10,000).
    - Advantage: Low memory overhead, uses Python's highly optimized C-level sorting.
    """
    # Calculate capacities and ball totals - O(n^2)
    row_sums = [sum(row) for row in container]
    col_sums = [sum(col) for col in zip(*container)]

    # Sort in-place to minimize memory allocation - O(n log n)
    row_sums.sort()
    col_sums.sort()

    # Direct list comparison - O(n)
    return "Possible" if row_sums == col_sums else "Impossible"

def organizing_containers_counter(container):
    """
    Approach 2: The Frequency Map Method (Hash Map)
    - Complexity: O(n) average case.
    - Best for: Large-scale datasets where n is very high.
    - Advantage: Linear scaling; avoids the overhead of ordering elements.
    """
    # Calculate capacities and ball totals - O(n^2)
    row_sums = [sum(row) for row in container]
    col_sums = [sum(col) for col in zip(*container)]

    # Compare distribution of values using Hash Maps - O(n)
    # This checks if the counts of each sum value match exactly.
    if Counter(row_sums) == Counter(col_sums):
        return "Possible"
    else:
        return "Impossible"

def main():
    q = int(input().strip())

    for q_itr in range(q):
        n = int(input().strip())

        container = []

        for _ in range(n):
            container.append(list(map(int, input().rstrip().split())))

        print(organizingContainers(container))
        print(organizing_containers_counter(container))

if __name__ == '__main__':
    main()